### PR TITLE
Environment config SPA fixes

### PR DESCRIPTION
**Pull request by vrushaliwaykole**
Issue: #7176 

Description:
1. Disabled delete button on new environments SPA if the environment is partially defined in config repo.
2. Rename method `editEnvironmentsPath` to `getEnvironmentPathOnSPA`
3. Disable remove environment variable button when it is defined in the config repository.
4. Fix for: Invalid agent count is shown when the environment contains remote agents. https://github.com/gocd/gocd/issues/7176#issuecomment-559349516
5. Fix for: Non-existing agent in the remote environment causes the addition of other agents to fail https://github.com/gocd/gocd/issues/7176#issuecomment-559353500

### CODE REVIEW

File: server/src/main/webapp/WEB-INF/rails/webpack/views/pages/new-environments/environments_widget.tsx [hunk 1]
```diff
@@ -79,12 +79,18 @@ export class EnvironmentWidget extends MithrilViewComponent<EnvAttrs> {
         </div>
       </div>;
     }
+    let deleteTitle;
+    if (!environment.canAdminister()) {
+      deleteTitle = `You are not authorized to delete the '${environment.name()}' environment.`;
+    } else if (!environment.isLocal()) {
+      deleteTitle = `Cannot delete '${environment.name()}' environment as it is partially defined in config repo.`;
```
Comment (GaneshSPatil) [inline server/src/main/webapp/WEB-INF/rails/webpack/views/pages/new-environments/environments_widget.tsx]: can we change `Cannot delete '${environment.name()}' environment as it is partially defined in config repository.`?

Changing `repo` => `repository`?

Comment (vrushaliwaykole) [reply server/src/main/webapp/WEB-INF/rails/webpack/views/pages/new-environments/environments_widget.tsx]: will change it.

File: server/src/main/webapp/WEB-INF/rails/webpack/views/components/environment_variables/spec/index_spec.tsx [hunk 2]
```diff
@@ -42,45 +53,46 @@ describe("Environment Variables Widget", () => {
 
     it("should have plain text environment variable fields", () => {
       const plainTextValue = plainTextEnvVar.value();
-      expect(helper.byTestId("env-var-name")).toHaveValue(plainTextEnvVar.name());
-      expect(helper.byTestId("env-var-value")).toHaveValue(plainTextValue ? plainTextValue : "");
-      expect(helper.byTestId("remove-env-var-btn")).toBeInDOM();
+      const wrapper        = helper.allByTestId("environment-variable-wrapper")[0];
+      expect(helper.byTestId("env-var-name", wrapper)).toHaveValue(plainTextEnvVar.name());
+      expect(helper.byTestId("env-var-value", wrapper)).toHaveValue(plainTextValue ? plainTextValue : "");
```
Comment (viraj2712) [inline server/src/main/webapp/WEB-INF/rails/webpack/views/components/environment_variables/spec/index_spec.tsx]: Should this be explicit assertion?
`if/else` in assertion doesn't make sense

File: server/src/main/webapp/WEB-INF/rails/webpack/views/pages/new-environments/spec/environments_widget_spec.tsx [hunk 3]
```diff
@@ -49,67 +43,99 @@ describe("Environments Widget", () => {
...
+    expect(deleteIcons).toHaveLength(2);
     expect(deleteIcons[0].title).toBe(`You are not authorized to delete the '${xmlEnv.name}' environment.`);
     expect(deleteIcons[0]).toBeDisabled();
     expect(deleteIcons[1].title).toBeFalsy();
     expect(deleteIcons[1]).not.toBeDisabled();
-    expect(deleteIcons[2].title).toBeFalsy();
-    expect(deleteIcons[2]).not.toBeDisabled();
+  });
+
+  it("should render disabled delete icon for the environments that user can not administer", () => {
```
Comment (viraj2712) [inline server/src/main/webapp/WEB-INF/rails/webpack/views/pages/new-environments/spec/environments_widget_spec.tsx]: How about `should render disabled delete icon for the environments defined in config repo and user can not administer`
as it verifies the same
https://github.com/gocd/gocd/pull/7320/files#diff-16caf4c51b6acb7dd2c1cb0bcadf106aR109

File: api/api-internal-environments-v1/src/main/java/com/thoughtworks/go/apiv1/internalenvironments/InternalEnvironmentsControllerV1.java [hunk 4]
```diff
@@ -108,17 +109,15 @@ public String indexMergedEnvironments(Request request, Response response) throws
...
-        Optional<List<String>> uuids = jsonReader.readStringArrayIfPresent("uuids");
-        if (uuids.isPresent()) {
-            return uuids.get();
+        JsonReader agents = jsonReader.readJsonObject("agents");
+
+        List<String> uuidsToAssociate = agents.readStringArrayIfPresent("add").orElse(Collections.emptyList());
+        List<String> uuidsToRemove = agents.readStringArrayIfPresent("remove").orElse(Collections.emptyList());
+
+        if (!uuidsToAssociate.isEmpty() || !uuidsToRemove.isEmpty()) {
+            agentService.updateAgentsAssociationOfEnvironment(envConfig, uuidsToAssociate, uuidsToRemove);
```
Comment (GaneshSPatil) [inline api/api-internal-environments-v1/src/main/java/com/thoughtworks/go/apiv1/internalenvironments/InternalEnvironmentsControllerV1.java]: @vrushaliwaykole @dhanasp -- There is a null check inside `updateAgentsAssociationOfEnvironment` as there is only one usage of this method and we're already making sure that the environment exists as part of `environmentConfigService.getEnvironmentConfig` should we remove the null check verification from `updateAgentsAssociationOfEnvironment` ?

File: api/api-internal-environments-v1/src/main/java/com/thoughtworks/go/apiv1/internalenvironments/InternalEnvironmentsControllerV1.java [hunk 5]
```diff
@@ -108,17 +109,15 @@ public String indexMergedEnvironments(Request request, Response response) throws
...
         }
-        throw halt(422, MessageJson.create(format("Json `%s` does not contain property '%s'.", new Gson().fromJson(request.body(), JsonElement.class), "uuids")));
+        return renderMessage(response, 200, "Environment '" + envName + "' updated successfully!");
```
Comment (GaneshSPatil) [inline api/api-internal-environments-v1/src/main/java/com/thoughtworks/go/apiv1/internalenvironments/InternalEnvironmentsControllerV1.java]: How about `return renderMessage(response, 200, EntityType.Environment("bob").updateSuccessful());` ?

We will need to define `updateSuccessful` utility method similar to `deleteSuccessful`

**Comment by GaneshSPatil** [comment]
#### Issues with internal Environments API update:

- [ ] Use PATCH as request method instead of PUT.
- [ ] Improve error message. Change `Agents with uuids 'agent-1,agent-2' were not found!` to `Agents with uuids ["agent-1", "agent-2"] were not found!`
- [ ] Removal of an agent associated in config repository should fail with Unprocessible Entity 422 error. Whereas the current API returns 200 with no change in the environment.
- [ ] Should the API fail when a non-associated agent is removed from the environment?
- [ ] Should the API fail when the already associated agent is added to the environment again?
- [ ] Should the API fail when the same agent is is in both `add` and `remove` list?


#### Open question:
- [ ] Should this be a public API?
- [ ] Should this API support bulk update for pipelines and environment variables too (in case it is a public API)?

**Comment by kritika-singh3** [comment]
> * [ ]  Improve error message. Change `Agents with uuids 'agent-1,agent-2' were not found!` to `Agents with uuids ["agent-1", "agent-2"] were not found!`

The error message is a generic error message generated via `EntityType#notFoundMessage` when a `RecordNotFoundException` is thrown. Do you still recommended the change?

> * [ ]  Removal of an agent associated in config repository should fail with Unprocessible Entity 422 error. Whereas the current API returns 200 with no change in the environment.
> * [ ]  Should the API fail when a non-associated agent is removed from the environment?
> * [ ]  Should the API fail when the already associated agent is added to the environment again?
> * [ ]  Should the API fail when the same agent is is in both `add` and `remove` list?

All these scenarios are not achievable via SPA. This internal API in question was created for SPA use only. Do we still need to think of fixing these scenarios?

File: api/api-internal-environments-v1/src/main/java/com/thoughtworks/go/apiv1/internalenvironments/InternalEnvironmentsControllerV1.java (hunk 4 above)
Comment (kritika-singh3) [reply api/api-internal-environments-v1/src/main/java/com/thoughtworks/go/apiv1/internalenvironments/InternalEnvironmentsControllerV1.java]: @GaneshSPatil agreed that the method `updateAgentsAssociationOfEnvironment` is currently only used in the Internal controller. But is that enough to remove the check.
The service can be called from anywhere (in future) and shouldn't it be the responsibility of the method itself to handle all the cases? WDYT?
cc:\ @viraj2712

Comment (GaneshSPatil) [reply api/api-internal-environments-v1/src/main/java/com/thoughtworks/go/apiv1/internalenvironments/InternalEnvironmentsControllerV1.java]: There should be no reason a `null` is sent for environment config. I couldn't think of any scenario where an environment is null, hence the thought of removing that check.

**Pull request merged**